lib/server/server.mjs

```diff
@@ -0,0 +1,406 @@
+import { InvalidMessageError, StreamingError } from './error.mjs'
+
+/**
+* @summary Request handler callback.
+*
+* @description Request handlers are callbacks responding to a single incoming
+* request by inspecting the `request` and using the `response` to stream data as
+* appropriate.
+*
+* A single request may go through multiple request handlers. To enable data
+* sharing between request handlers, a `context` object is provided that can be
+* freely modified.
+*
+* For example, the first handler may take the `authorization` string from the
+* message header, look up the corresponding user from the DB and store the data
+* in the `context` object. Then, the next handler can perform operations on the
+* user, without needing to know how the user was authenticated.
+*
+* @callback RequestHandler
+* @param {Request} request Request data
+* @param {Response} response Response instance
+* @param {object} context Context object
+* @typedef {function(Request, Response, object)} RequestHandler
+*/
+
+/**
+* @summary Request exception handler callback.
+*
+* @description Exception handlers are called in cases when an exception is
+* thrown during request processing. Exception handlers act similar to regular
+* exception handlers, in the sense that a single request may go through multiple
+* exception handlers, and they may stream response data in return.
+*
+* They also have access to the same data as the regular request handlers, in
+* addition to the exception that occurred during processing.
+*
+* @callback RequestExceptionHandler
+* @param {Request} request Request data
+* @param {Response} response Response instance
+* @param {object} context Context object
+* @param {any} exception Exception occurred
+* @typedef {function(Request, Response, object, any)} RequestExceptionHandler
+*/
+
+/**
+* @summary Default request handler.
+*
+* @description This handler should be registered as a default handler, in cases
+* where no handlers are registered for the message subject. The response will
+* always be the following:
+*
+* ```js
+* {
+*   "header": {
+*     // original header
+*   },
+*
+*   "type": "err",
+*
+*   "error": {
+*     "type": "UnknownSubject",
+*     "message": "Unknown subject: <subject>"
+*   }
+* }
+* ```
+*
+* @param {Request} request Request data
+* @param {Response} response Response instance
+* @param {object} _context Context object
+*/
+export function unknownSubjectHandler (request, response, _context) {
+  response.error(new MessageError({
+    type: 'UnknownSubject',
+    message: `Unknown subject: ${request.header.subject}`
+  }))
+}
+
+/**
+* @summary Default exception handler.
+*
+* @description This exception handler is always registered and serves as a
+* catch-all exception handler. It will try to extract the exception's name and
+* message to be used as error type and message. Otherwise it will fall back to a
+* generic 'UnknownError' without message.
+*
+* @param {Request} _request Request data
```

Comment:
Why are the request, response and context parameters are marked as "logically private" here? The method on line #82 does the same for the context object only (and I still don't get it).

Reply:
They are prefixed with an underscore to show that they are unused. I didn't want to leave them out and/or use a plain underscore because I wanted to leave in info about what we're not using in the particular method.

Reply:
Aaand again, missing knowledge in JS ecosystem, thanks for the clarification! :)